driver-core/src/main/java/com/datastax/driver/core/CodecRegistry.java

```diff
@@ -219,7 +201,66 @@ public int hashCode() {
         }
     }
 
-    private final boolean immutable;
+    /**
+     * A complexity-based weigher.
+     * Weights are computed according to the CQL type:
+     * <ol>
+     * <li>Manually-registered codecs always weigh 0;
+     * <li>Codecs for primtive types weigh 0;
+     * <li>Codecs for collections weigh the total weight of their inner types + the weight of their level of deepness;
+     * <li>Codecs for UDTs and tuples weigh the total weight of their inner types + the weight of their level of deepness, but cannot weigh less than 1;
+     * <li>Codecs for custom (non-CQL) types weigh 1.
+     * </ol>
+     * A consequence of this algorithm is that primitive types and all "shallow" collections thereof
+     * are never evicted.
+     */
+    private class TypeCodecWeigher implements Weigher<CacheKey, TypeCodec<?>> {
+
+        @Override
+        public int weigh(CacheKey key, TypeCodec<?> value) {
+            return codecs.contains(value) ? 0 : weigh(key.cqlType, 0);
+        }
+
+        private int weigh(DataType cqlType, int level) {
+            switch (cqlType.getName()) {
+                default:
+                    return 0;
```

Comment:
I found this really confusing. I know it works (TBH I wrote a test to verify it), but it's going to give pause to anyone reading the code. For the sake of conventions, default clause should go last.